faidx.c

```diff
@@ -42,18 +42,76 @@ DEALINGS IN THE SOFTWARE.
 #include <limits.h>
 #include <htslib/faidx.h>
 #include <htslib/hts.h>
+#include <htslib/hfile.h>
+#include <htslib/kstring.h>
 #include "samtools.h"
 
 #define DEFAULT_FASTA_LINE_LEN 60
 
+
+static int write_fasta(faidx_t *faid, FILE *file, const char *name, const int ignore, const int length) {
+    int seq_len, beg, end;
+    char *seq = fai_fetch(faid, name, &seq_len);
+
+    fprintf(file, ">%s\n", name);
+
+    if ( seq_len < 0 ) {
+        fprintf(stderr, "[faidx] Failed to fetch sequence in %s\n", name);
+
+        if (ignore && seq_len == -2) {
+            return EXIT_SUCCESS;
+        } else {
+            return EXIT_FAILURE;
+        }
+    } else if (seq_len == 0) {
+        fprintf(stderr, "[faidx] Zero length sequence: %s\n", name);
+    } else if (hts_parse_reg(name, &beg, &end) && (end < INT_MAX) && (seq_len != end - beg)) {
+        fprintf(stderr, "[faidx] Truncated sequence: %s\n", name);
+    }
+
+    size_t i, seq_sz = seq_len;
+
+    for (i=0; i<seq_sz; i+=length)
+    {
+        size_t len = i + length < seq_sz ? length : seq_sz - i;
+        if (fwrite(seq + i, 1, len, file) < len ||
+            fputc('\n', file) == EOF) {
+            print_error_errno("faidx", "failed to write output");
+            return EXIT_FAILURE;
+        }
+    }
+
+    free(seq);
+
+    return EXIT_SUCCESS;
+}
+
+
+static int read_regions_from_file(faidx_t *faid, hFILE *in_file, FILE *file, const int ignore, const int length) {
+    kstring_t line = {0, 0, NULL};
+    int ret = EXIT_FAILURE;
+
+    while (line.l = 0, kgetline(&line, (kgets_func *)hgets, in_file) >= 0) {
+        if ((ret = write_fasta(faid, file, line.s, ignore, length)) == EXIT_FAILURE) {
+            break;
+        }
+    }
+
+    free(line.s);
+
+    return ret;
+}
+
+
 static int usage(FILE *fp, int exit_status)
 {
     fprintf(fp, "Usage: samtools faidx <file.fa|file.fa.gz> [<reg> [...]]\n");
     fprintf(fp, "Option: \n"
-                " -o, --output   FILE Write FASTA to file.\n"
-                " -n, --length   INT  Length of FASTA sequence line. [60]\n"
-                " -c, --continue      Continue after trying to retrieve missing region.\n"
-                " -h, --help          This message.\n");
+                " -o, --output      FILE Write FASTA to file.\n"
+                " -n, --length      INT  Length of FASTA sequence line. [60]\n"
+                " -c, --continue         Continue after trying to retrieve missing region.\n"
+                " -r, --region_file FILE File of regions.  One per line.\n"
```

Comment:
It should be stated that a line in the region file has the format **reg:from-to** (`chr1:12209228-12209246`) and it is **1-based**, so to be clear that is not a BED file or any other TAB delimited format.

Reply:
Is there a reason it couldn't additionally accept the regions as a BED file? This would be useful and in line with other samtools commands (though starting to duplicate the already-existing `bedtools getfasta`).

While you're editing the usage… wouldn't putting FILE/INT/etc beside their options be more readable?

    " -o, --output FILE       Write FASTA to file\n"
    …etc…
    " -r, --region_file FILE  File of regions.  One per line\n"

Reply:
Oh, also usual would be `--region-file` with hyphen rather than underscore.

Reply:
Regions have their own section in the man page and neither the current faidx usage or man entry defines them.  I will add a little bit more to the usage but not a full definition.  BED files are a different issue and if anyone wants to raise one then I will look at it.  Yes it should be --region-file.